### PR TITLE
v0.8.3 - Fix unpairing crash

### PR DESCRIPTION
**Pull request by JeremyChiang**
Fixes: #198 

@kylebrowning could you also review this PR? Thanks 👍🏻

## Summary
This fix allows Bluejay to complete the disconnection clean up without crashing when a device is unpaired.

## Solution
I've never triggered iOS' system pairing dialog before with Bluejay, because I've never worked with a characteristic that is encrypted, since most of our client projects deal with authentication upfront, not on per-characteristic basis.

But I agree this can be common, so this fix is important.

Anyhow, I didn't know we can also get a disconnection event while the connection operation is in the `running` state, which is exactly what happens when:
1. Bluejay is attempting to reconnect
2. If you unpair the device during a reconnect attempt

This fix simply adds that specific case, and provide the missing `connectingError` using `BluejayError.unexpectedDisconnect`

```
if wasConnecting {
    debugLog("Disconnect clean up: delivering expected disconnected event back to the pending connection in the queue...")

    if let connection = weakSelf.queue.first as? Connection {
        if case .running = connection.state {
            connectingError = BluejayError.unexpectedDisconnect
        } else if case let .stopping(error) = connection.state {
            connectingError = error
        }
    }

}
```

## Notes
- I've added basic ways to test pairing and unpairing in our demo apps
 - You must accept pairing on both BluejayDemo and DittojayDemo to complete the pairing process
- You cannot test iOS pairing using the demos if both of your devices are logged into the same iCloud account (surprise 🤭)
- This fix might also cover some other unknown crashes related to unexpected disconnections during a connection
- This fix may not cover all crashes related to pairing and unpairing

### CODE REVIEW

**Comment by JeremyChiang** [comment]
## Extra Thoughts 
(I may put this in the Wiki after some revisions and fact-checks)

After some research and experimentation, "pairing" in BLE can be a cluster[insert your word of choice] for many reasons.

In general, the concept of "pairing" doesn't really apply to BLE. Any "pairing" functionality is usually custom at the firmware and application logic level, not at the Bluetooth LE level - for the benefit of general users who are familiar with Bluetooth classic, but don't know and don't care about the difference between classic vs LE.

For example, once a Bluetooth classic device is paired, (I believe) the classic protocol has some smarts and reconnection logic already built-in. In BLE, and in Bluejay as well, auto-reconnect is only possible through a manual implementation that is not part of the Bluetooth LE's protocol - something we have to add ourselves (to make it behave like classic...sort of).

On a high level, the primary communication concepts that actually apply to BLE are broadcasting, advertising, and connection, so "pairing" is technically not a part of the BLE protocol. Except to make matter worse, the negotiation of broadcasting, advertising, and connection in BLE, are also sometimes called "pairing", which is a little different from the "pairing" that we'd normally refer to and associate with classic 😅

Then as if that's not enough, the pairing dialog you see in iOS when working with BLE, is actually more correctly called "bonding" that is **also** not the same as the "pairing" you and I tend to think of, where once iOS completes the "pairing/bonding", the clients can then freely read and write to characteristics that are flagged as encrypted in the GATT profile, **but does not actually reconnect without something like Bluejay**. And from what I've tried and read, there's no way to tell iOS to not show this pairing dialog the first time you try to read or write an encrypted characteristic.

What this means is your app could potentially have 2 sets of pairing process and UI if you are not careful - and that's bad for UX. Or you could end up with a custom pairing logic, but not have any devices to unpair in iOS' Bluetooth settings if you've never triggered "bonding".

In summary, if you need "pairing" for your BLE device, always do that through or include iOS' pairing (bonding) dialog along with your custom pairing requirements (setting or exchanging certain values and defaults...etc), otherwise, don't use the term "pair" in any of your authentication flow (because there will be no devices to unpair in Bluetooth settings), and don't have any "rogue" characteristics that are encrypted at the GATT level, which could trigger an unexpected pairing dialog.

**Comment by kylebrowning** [comment]
👍 Crash no longer happens.
